python/tvm/autodiff.py

```diff
@@ -0,0 +1,130 @@
+"""Namespace of autodiff-related functions.
+
+The functions are automatically exported from C++ side via PackedFunc.
+You can read "include/tvm/autodiff.h" for the function signature of these functions.
+"""
+import logging
+
+from ._ffi.function import _init_api
+from ._ffi.node import NodeBase, register_node
+
+_init_api("tvm.autodiff")
+
+@register_node
+class DifferentiationResult(NodeBase):
+    """Result of differentiation.
+
+    Parameters
+    ----------
+    result : list of Tensor
+        The requested adjoints, i.e. the jacobians or gradients of the given output
+        wrt to the given inputs.
+
+    adjoints : dict from Tensor to Tensor
+        A map from tensors to the corresponding adjoints (including internal nodes).
+
+    adjoint_summands : dict from Tensor to dict from Tensor to Tensor
+        Single summands of the adjoints.
+    """
+    def __getattr__(self, name):
+        # Here we convert tvm Maps to dicts because Map compares keys by reference which is
+        # wrong for Tensors. Hopefully, in the future Map gets fixed somehow, and this function
+        # may be removed then.
+        res = NodeBase.__getattr__(self, name)
+        if name == 'adjoints':
+            return dict(res.items())
+        if name == 'adjoint_summands':
+            return {k: dict(v.items()) for k, v in res.items()}
+        return res
+
+    def __getitem__(self, i):
+        return self.result[i]
+
+    def __len__(self):
+        return len(self.result)
+
+
+def differentiate(output, inputs=None, head=None, manual=None, fdiff=None):
+    """Perform reverse-mode automatic differentiation.
+
+    Example::
+
+        x = tvm.placeholder((32, 3, 28, 28), name='x')
+        w1 = tvm.placeholder((10, 3, 3, 3), name='w1')
+        w2 = tvm.placeholder((10, 10, 3, 3), name='w2')
+        y = topi.sum(topi.nn.conv2d(topi.nn.conv2d(x, w1, 1, 0), w2, 1, 0))
+
+        [dw1, dw2] = tvm.differentiate(y, [w1, w2])
+
+    Parameters
+    ----------
+    output : Tensor
+        The tensor to differentiate.
+
+    inputs : list of Tensor
+        The list of input tensors. When the list is empty or None, will perform
+        differentiation wrt all tensors the output depends on (i.e. will compute all
+        adjoints and populate the corresponding dict, but the list of results
+        will be empty).
+
+    head : Tensor
```

Comment:
Can this be understood as the gradient of the output tensor w.r.t. the final output of a math expression? Could you explain why its shape is `output.shape+output.shape` when `None` is passed? My understanding is that it should default to `ones(output.shape)`. Thanks.

Reply:
Please disregard my question above. After reading your implementation of `DiffBuildingBlock`, I think I got the meaning of defaulting `head` to shape `output.shape+output.shape`.